File: src/main/java/xyz/phanta/cbtweaker/hatch/HatchBlockItem.java
package xyz.phanta.cbtweaker.hatch;

import io.github.phantamanta44.libnine.client.model.ParameterizedItemModel;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraft.world.World;
import xyz.phanta.cbtweaker.util.item.CustomModelItem;

import javax.annotation.Nullable;
import java.util.Objects;

public class HatchBlockItem extends ItemBlock implements CustomModelItem, ParameterizedItemModel.IContextSensitive {

    private final HatchBlock block;

    public HatchBlockItem(HatchBlock block) {
        super(block);
        this.block = block;
        setHasSubtypes(true);
    }

    @Override
    public HatchBlock getBlock() {
        return block;
    }

    @Override
    public int getMetadata(int damage) {
        return damage;
    }

    @Override
    public String getTranslationKey(ItemStack stack) {
        return block.getTranslationKey() + "." + stack.getMetadata();
    }

    @Override
    public void collectItemModels(ModelAcceptor modelAcceptor) {
        String path = Objects.requireNonNull(getRegistryName()).getPath();
        for (int i = 0; i < block.getHatchType().getTierCount(); i++) {
            modelAcceptor.accept(i, path);
        }
    }

    @Override
    public void getModelMutations(ItemStack stack, @Nullable World world, @Nullable EntityLivingBase holder,
                                  ParameterizedItemModel.Mutation m) {
        m.mutate("tier", Integer.toString(stack.getMetadata()));
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/hatch/HatchBlockItem.kt
package st.evening.mc.cbtweaker.hatch

import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraft.item.ItemStack
import net.minecraftforge.client.model.ModelLoader
import st.evening.mc.cbtweaker.common.CbtCustomBlockItem
import st.evening.mc.prelude.api.item.CustomModelItem
import st.evening.mc.prelude.api.util.game.ClientSide
import st.evening.mc.prelude.api.util.game.mapPath

class HatchBlockItem(private val hatchBlock: HatchBlock) : CbtCustomBlockItem(hatchBlock), CustomModelItem {
    init {
        setHasSubtypes(true)
    }

    override fun getMetadata(damage: Int): Int = damage

    override fun getTranslationKey(stack: ItemStack): String = hatchBlock.translationKey + "." + stack.metadata

    @ClientSide.Physical
    override fun registerItemModels() {
        val regName = registryName!!
        for (i in 0..<hatchBlock.hatchType.tierCount) {
            ModelLoader.setCustomModelResourceLocation(
                this, i, ModelResourceLocation(regName.mapPath { "$it/${it}_$i" }, "inventory")
            )
        }
    }
}
